src/commands/signzone.rs

```diff
@@ -0,0 +1,4157 @@
+            Timestamp::from_str(arg)
+        };
+
+        res.map_err(|err| Error::from(format!("Invalid timestamp: {err}")))
+    }
+
+    pub fn execute(self, env: impl Env) -> Result<(), Error> {
+        // Post-process arguments.
+        // TODO: Can Clap do this for us?
+
+        let signing_mode = if self.hash_only {
+            SigningMode::HashOnly
+        } else if self.key_paths.is_empty() {
+            if self.allow_zonemd_without_signing {
+                SigningMode::None
+            } else {
+                return Err("Missing key argument".into());
+            }
+        } else {
+            SigningMode::HashAndSign
+        };
+
+        let out_file = if let Some(out_file) = &self.out_file {
+            out_file.clone()
+        } else {
+            let out_file = format!("{}.signed", self.zonefile_path.display());
+            PathBuf::from_str(&out_file)
+                .map_err(|err| format!("Cannot write to {out_file}: {err}"))?
+        };
+
+        // ldns-signzone only shows these warnings if verbosity < 1 but offers
+        // no way to configure the verbosity level. I assume the intent was to
+        // add support for a -q (--quiet) option or similar but that was never
+        // done.
+        match self.iterations {
+            500.. => Self::write_extreme_iterations_warning(&env),
+            100.. if self.invoked_as_ldns => Self::write_large_iterations_warning(&env),
+            1.. if !self.invoked_as_ldns => Self::write_non_zero_iterations_warning(&env),
+            _ => { /* Good, nothing to warn about */ }
+        }
+
+        // Read the zone file.
+        let mut records = self.load_zone(&env.in_cwd(&self.zonefile_path))?;
+
+        // Extract the SOA RR from the loaded zone.
+        let Some(soa_rr) = records.find_soa() else {
+            return Err(format!(
+                "Zone file '{}' does not contain a SOA record",
+                self.zonefile_path.display()
+            )
+            .into());
+        };
+        let ZoneRecordData::Soa(_) = soa_rr.first().data() else {
+            return Err(format!(
+                "Zone file '{}' contains an invalid SOA record",
+                self.zonefile_path.display()
+            )
+            .into());
+        };
+
+        let dnskey_rrset = records.find_apex_rtype(soa_rr.owner(), Rtype::DNSKEY);
+
+        let cds_rrset = records.find_apex_rtype(soa_rr.owner(), Rtype::CDS);
+        let cdnskey_rrset = records.find_apex_rtype(soa_rr.owner(), Rtype::CDNSKEY);
+
+        // Extract and validate the DNSKEY RRs from the loaded zone.
+        let mut found_public_keys = vec![];
+        for rr in records.iter() {
+            if let ZoneRecordData::Dnskey(dnskey) = rr.data() {
+                // Create a public key object from the found DNSKEY RR.
+                let public_key = Record::new(rr.owner(), Class::IN, Ttl::ZERO, dnskey);
+
+                found_public_keys.push(public_key);
+            }
+        }
```

Comment:
I think this code will iterate over all records in the zone looking for DNSKEYs, yet they were already found at line 659 above, so just iterate over those instead.... unless I'm missing something.

```suggestion
        if let Some(dnskey_rrset) = &dnskey_rrset {
            for rr in dnskey_rrset.iter() {
                if let ZoneRecordData::Dnskey(dnskey) = rr.data() {
                    // Create a public key object from the found DNSKEY RR.
                    let public_key = Record::new(rr.owner(), Class::IN, Ttl::ZERO, dnskey);

                    found_public_keys.push(public_key);
                }
            }
        }
```

Reply:
@Philip-NLnetLabs: Any opinion on this?

Reply:
I think this change is correct.